### ISSUE TITLE
Search engine Data model for Parks - Analysis

### ISSUE THREAD

**Issue by KulturExplora** [KulturExplora/toursearchengine#37]
Analyse the possible open source libraries that could be used to build the machine learning algorithm for searching the national park names within the Tour description, Itinerary, other columns, etc.

**Comment by khiba-k** [comment]
#  Gemini AI API Analysis: Mapping Safari Tours to National Parks

##  Objective

The goal of this test was to evaluate whether the **free Gemini AI API** can correctly identify and associate **national parks** with given **safari tour names**, based solely on subtle hints in the title.

---

##  Test Data

I provided the model with two lists:

### Parks

[Addo Elephant National Park, South Africa,  
Masai Mara National Reserve, Kenya,  
Okavango Delta, Botswana,  
Serengeti National Park, Tanzania,  
Ngorongoro Conservation Area, Tanzania,  
Kruger National Park, South Africa,  
Chobe National Park, Botswana]

###  Safari Tours

[7 Days High End Crater & Culture Circuit,  
9 Days High End Baobab & Bushveld Safari,  
8 Days High End Great Herds Migration Journey,  
10 Days High End Mokoro Water Safari,  
6 Days High End Big Five Luxe Escape , 
7 Days High End Salt Pan & Sunset Trail,  
11 Days High End Predator Plains Adventure,  
8 Days High End River of Giants Safari,  
9 Days High End Flamingo Lakes Circuit,  
7 Days High End Savannah Giants Experience]

---

## Gemini Output

The Gemini model returned the following matches:

'tourName: 7 Days High End Crater & Culture Circuit, park: Ngorongoro Conservation Area, Tanzania'  
'tourName: 9 Days High End Baobab & Bushveld Safari, park: Kruger National Park, South Africa'  
'tourName: 8 Days High End Great Herds Migration Journey, park: Masai Mara National Reserve, Kenya'  
'tourName: 10 Days High End Mokoro Water Safari, park: Okavango Delta, Botswana'  
'tourName: 6 Days High End Big Five Luxe Escape, park: Kruger National Park, South Africa'  
'tourName: 7 Days High End Salt Pan & Sunset Trail, park: Makgadikgadi Pans National Park, Botswana'  
'tourName: 11 Days High End Predator Plains Adventure, park: Serengeti National Park, Tanzania'  
'tourName: 8 Days High End River of Giants Safari, park: Chobe National Park, Botswana'  
'tourName: 9 Days High End Flamingo Lakes Circuit, park: Lake Nakuru National Park, Kenya'

---

##  Billing Information - model_name="gemini-1.5-flash"

### Free of charge

#### RATE LIMITS
- 15 RPM (Requests Per Minute)
- 1,000,000 TPM (Total Requests Per Month)
- 1,500 RPD (Requests Per Day)

#### PRICE (INPUT)
- Free of charge

#### PRICE (OUTPUT)
- Free of charge

#### GROUNDING WITH GOOGLE SEARCH
- Not available

#### DATA USED TO IMPROVE  PRODUCT
- Yes

**Comment by AYSamuel** [comment]
To build a machine learning algorithm that can identify and associate national park names within safari tour descriptions, itineraries, and other text columns, several open-source libraries are well-suited for natural language processing (NLP) and text classification tasks. These libraries can help extract named entities (such as park names), perform text classification, and handle subtle semantic hints in tour titles.

**Key Open Source Libraries for NLP and Text Classification**

**1. spaCy**

spaCy is a fast, efficient, and production-ready NLP library in Python.
It provides pre-trained models for named entity recognition (NER), which can identify location names like national parks in text.
It supports tokenization, part-of-speech tagging, dependency parsing, and custom entity recognition training if needed.
spaCy is suitable for processing large volumes of text quickly and can be extended with custom pipelines to improve recognition of domain-specific terms like park names.

**2. NLTK (Natural Language Toolkit)**
NLTK is a comprehensive Python library for NLP with a wide range of tools for tokenization, parsing, classification, and named entity recognition.
It is beginner-friendly and has extensive documentation and community support.
NLTK can be used to preprocess text and implement classical machine learning models such as Naive Bayes classifiers for text classification tasks.

**3. Gensim**
Gensim specializes in topic modeling and semantic similarity analysis.
It can be used to extract latent topics or semantic relationships from large text corpora, which might help in associating subtle hints in tour descriptions with specific parks.
Gensim is efficient for handling large datasets and can complement entity recognition by providing context-based similarity measures.

**4. Scikit-learn**
Scikit-learn offers a broad range of machine learning algorithms for classification, clustering, and regression.
It can be used to build classifiers (e.g., SVM, random forest, Naive Bayes) that categorize text snippets into park-related labels based on features extracted from the text.

Useful for feature engineering and model evaluation in text classification pipelines.

**5. Apache OpenNLP and MALLET (Java-based)**

For Java environments, Apache OpenNLP provides tools for tokenization, sentence segmentation, and named entity extraction.
MALLET is focused on text classification and topic modeling and supports advanced algorithms for text categorization and entity extraction.

**Suggested Approach for the Safari Tour Use Case**

**Named Entity Recognition (NER)**: Use spaCy or NLTK to detect and extract mentions of national parks or related geographic entities in tour descriptions and titles. Custom training may be required to improve recognition of specific park names.

**Text Classification**: Use Scikit-learn or Naive Bayes classifiers to categorize tours based on textual features into park-related classes, especially when park names are implied rather than explicitly mentioned.

**Semantic Similarity and Topic Modeling:** Use Gensim to analyze the semantic context of tour descriptions to associate them with parks based on latent topics or similarity to known park-related keywords.

**Data Preprocessing**: Tokenization, lemmatization, and stop-word removal can be done via spaCy or NLTK to prepare text for modeling.

**Comment by KulturExplora** [comment]
List of safari tours. Each tour has details like its title, description, daily itinerary, and what's included or excluded.

Create a list of national parks in Tanzania and note the common names or keywords people use when referring to them (like "Serengeti" for Serengeti National Park, or "Ngorongoro" for Ngorongoro Crater).

For each tour, scan all its information (title, description, itinerary, etc.) and check if any of those park names or keywords are mentioned.

If a park is mentioned, tag that tour as covering that national park.

++++++++++++++++++++==========================
Below is the working python code
+++++++++++++++++++++++++=========================

import pandas as pd

# Load the CSV file
file_path = "Tours.csv"
df = pd.read_csv(file_path)

# Define the columns to search
text_columns = ["title", "description", "itinerary", "safariType", "included", "excluded", "country"]

# Define national parks and their keywords
safari_parks = {
    "Serengeti National Park": ["serengeti"],
    "Ngorongoro Conservation Area": ["ngorongoro"],
    "Lake Manyara National Park": ["manyara"],
    "Tarangire National Park": ["tarangire"],
    "Arusha National Park": ["arusha"],
    "Ruaha National Park": ["ruaha"],
    "Nyerere National Park": ["nyerere", "selous"],
    "Katavi National Park": ["katavi"],
    "Mahale Mountains National Park": ["mahale"],
    "Gombe Stream National Park": ["gombe"],
    "Mikumi National Park": ["mikumi"],
    "Udzungwa Mountains National Park": ["udzungwa"],
    "Saadani National Park": ["saadani"]
}

# Function to determine which parks are mentioned
def find_parks(row):
    text = " ".join(str(row[col]).lower() for col in text_columns if pd.notnull(row[col]))
    matched = [park for park, keywords in safari_parks.items() if any(k in text for k in keywords)]
    return ", ".join(matched)

# Apply the function to each row
df["covered_parks"] = df.apply(find_parks, axis=1)

# Show results: tour title + covered parks
for index, row in df.iterrows():
    print(f"{row['title']}: {row['covered_parks']}")

**Comment by khiba-k** [comment]
@KulturExplora, please share current status
